servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/PublisherBuffer.java

```diff
@@ -0,0 +1,353 @@
+            private final Subscriber<? super B> target;
+            private final Cancellable tCancellable;
+
+            BoundariesSubscriber(final Subscriber<? super B> target, final Cancellable tCancellable) {
+                this.target = target;
+                this.tCancellable = tCancellable;
+            }
+
+            @Override
+            public void onSubscribe(final Subscription subscription) {
+                ConcurrentSubscription bs = ConcurrentSubscription.wrap(new Subscription() {
+                    @Override
+                    public void request(final long n) {
+                        if (!isRequestNValid(n)) {
+                            // This code is deliberately not intending to avoid potentially concurrent calls to the
+                            // subscriber as the specification is already violated. This avoids complicating this
+                            // state machine for unexpected specification violations.
+                            Object oldUpdater = maybeCollectorUpdater.getAndSet(BufferSubscriber.this, TERMINATED);
+                            if (oldUpdater != TERMINATED) {
+                                subscription.cancel();
+                                safeOnError(target, newExceptionForInvalidRequestN(n));
+                            }
+                            return;
+                        }
+                        if (pendingBuffersUpdater.getAndAccumulate(BufferSubscriber.this, n,
+                                FlowControlUtils::addWithOverflowProtection) == 0) {
+                            Object cMaybeCollector = maybeCollector;
+                            if (cMaybeCollector instanceof DelayedTermination) {
+                                tryTerminate((DelayedTermination) cMaybeCollector);
+                                // as we are in the process of terminating do not send this request-n upstream.
+                                return;
+                            }
+                        }
+                        subscription.request(n);
+                    }
+
+                    @Override
+                    public void cancel() {
+                        try {
+                            subscription.cancel();
+                        } finally {
+                            tCancellable.cancel();
+                        }
+                    }
+                });
+                delayedCancellable(bs);
+                target.onSubscribe(bs);
+            }
+
+            @Override
+            public void onNext(@Nullable final Collector<T, B> collector) {
+                for (;;) {
+                    final Object cMaybeCollector = maybeCollector;
+                    if (cMaybeCollector == ADDING || cMaybeCollector == TERMINATED
+                            || cMaybeCollector instanceof DelayedTermination) {
+                        // If we are ADDING then we will try emitting in the next cycle
```

Comment:
`Publisher#buffer(..)` exposes a `Publisher<?>` for arbitrary signals so this could be driven by any type of event (maybe even an even that fires only upon certain amount of data that is produced which would make dropping `Collector` problematic). Even if the events are periodic the "next cycle" may not come for some time, or as mentioned you may get unlucky and have data be synchronized with timer (e.g. imagine lots of data being delivered). It would provide more predictable behavior to hand-off the latest `Collector` to ensure some data gets delivered as close to the boundary as possible, and it would be difficult avoid/understand these types of concurrency issues.

Reply:
@Scottmitch here the case is that the item is emitted concurrently with a boundary. 
Since, item is concurrently emitted, users should not have any expectation of which boundary the item should belong. We have the flexibility to determine whether the item should belong to the previous boundary or next. This code is making the decision of adding the item to the next boundary and handing-off the emission to the boundary source, which happens on the next boundary emission.